### ISSUE TITLE
Mas question

### ISSUE THREAD

**Issue by Holmes2002** [lswhim/Inductive-Cognitive-Diagnosis-Model#3]
Hi, when I was infering MAS for ED-Net 1 dataset, MAS values was all in around 0.4-0.6 ??? I thought proficiency of concepts should be ~ 0.8-0.9 in some cases? This MAR got from Eval function when training
```
[0.49719911 0.59806077 0.53006339 0.53206997 0.50294676 0.52654466
  0.44678597 0.46174606 0.47397813 0.49145323 0.53129938 0.51411308
  0.48975917 0.50262233 0.49328476 0.50180561 0.489658   0.52162171
  0.53231689 0.49229362 0.50165182 0.50804461 0.55280965 0.56324014
  0.52786051 0.54421552 0.54704067 0.55797429 0.49363904 0.49502383
  0.47689304 0.49737268 0.49903013 0.51700145 0.56841244 0.4983806
  0.49684432 0.49912599 0.54168052 0.55612338 0.49264415 0.49280173
  0.50695716 0.52135805 0.50382169 0.52798137 0.48751197 0.52547323
  0.50705572 0.52162952 0.49918278 0.49319787 0.49641511 0.4720429
  0.51115057 0.5536061  0.51384661 0.57996653 0.59803409 0.59643649
  0.51176925 0.48799572 0.501572   0.53766736 0.47806032 0.50863114
  0.49959809 0.46208058 0.4893647  0.51028475 0.50091929 0.53253236
  0.51688252 0.49046587 0.50611359 0.45753074 0.53697641 0.50769052
  0.5000894  0.50627527 0.49751146 0.48863612 0.50061136 0.49655883
  0.51729038 0.50004453 0.50353222 0.4979587  0.50538943 0.52121782
  0.48550864 0.50024026 0.50905837 0.51296533 0.5020818  0.50877185
  0.45686133 0.50943517 0.49729948 0.49873587 0.49365438 0.49800431
  0.49744102 0.48346997 0.49880125 0.48567211 0.50103792 0.51474721
  0.51950789 0.49572628 0.51161821 0.52570267 0.51880924 0.48036509
  0.50062999 0.52741683 0.50172769 0.47659103 0.4940972  0.52579701
  0.43690693 0.49740458 0.48163456 0.49981089 0.49970397 0.49516529
  0.50103011 0.51748771 0.48081865 0.47162139 0.50534043 0.5035737
  0.49265009 0.50050769 0.50771978 0.50011517 0.49936571 0.4980686
  0.49182935 0.49903657 0.49919903 0.49643831 0.47354603 0.46807897
  0.49925734 0.51128786 0.50145062 0.54902686 0.50211823 0.49444831
  0.56053411 0.54282303 0.4921654  0.4883848  0.58951139 0.49976814
  0.48165775 0.5063404  0.4994553  0.4844976  0.49836284 0.50208227
  0.49951317 0.43866058 0.49860004 0.49763996 0.51257903 0.59145758
  0.49770119 0.48818281 0.49243433 0.48236148 0.50397066 0.53172404
  0.54325278 0.60104938 0.50252715 0.62165399 0.50743004 0.53978417
  0.5335537  0.58560243 0.49569639 0.52990764 0.50805828 0.51326769
  0.52622047 0.49809032 0.55438333]]
```

**Comment by lswhim** [comment]
Hi, I understand that by MAR, you're referring to the Mas (mastery levels) defined in the paper, which estimate a student’s proficiency over concepts. This is actually a well-known limitation shared by many current cognitive diagnosis models. Since the final mastery levels are typically constrained within the range [0,1] using a sigmoid function, the outputs often cluster around 0.4–0.6—especially under sparse data conditions, such as when each student only attempts a small subset of the total question pool. Of course, other factors may also contribute to this behavior.

Our recent work, accepted at KDD 2024, investigates this issue in depth and proposes a method to alleviate it. If you're interested, feel free to check out our open-source implementation here: https://github.com/ECNU-ILOG/ORCDF.

**Comment by Holmes2002** [comment]
@lswhim  Oh, In case of real platform that how to estimate proficiency of a new student ? I have to retrain everytime when having a new student if yes how long it takes ?

**Comment by lswhim** [comment]
Great question! This is exactly the challenge addressed in our ICDM paper. Instead of retraining the model every time a new student joins, we construct a student-centric graph during training. In this graph, each existing student is connected to multiple building blocks—such as questions and knowledge concepts.

During training, we learn two key components:

The representations of these building blocks (e.g., questions and concepts), and

A learnable aggregation function that determines how to combine the features of the connected blocks.

So when a new student arrives, we don’t need to retrain the model. We simply plug the student into the graph based on their response history, and use the aggregation function to infer their mastery levels in an inductive manner. This makes the system highly efficient and scalable for real-world platforms.

**Comment by Holmes2002** [comment]
@lswhim Yes, I already knew the information because I had read the paper. I'm just a beginner in Edu-RecSys, and I want to ask: in what cases should CDF models like our recent work ORCDF be used? It seems difficult to use and retrain them continuously when the system constantly has new students

**Comment by Holmes2002** [comment]
Hm i received your response in ORCDF repo. Thanks for your support !

**Comment by lswhim** [comment]
Thanks for your interest — that’s a great question!

ORCDF was specifically designed to address the issue you mentioned earlier, where predicted mastery levels often cluster around 0.4–0.6. In our paper, we refer to this as the oversmoothing problem, and ORCDF introduces a response-aware graph structure to effectively alleviate it.

However, you're right that ORCDF, like many existing CDM models, assumes a static user set, meaning it’s mainly intended for standard settings where the student population is fixed and new users are not constantly joining the system.

On the other hand, our ICDM model was developed to handle the inductive setting, where new students (and potentially new items) can appear without retraining. It allows for dynamic reasoning over newly arriving users based on their response histories.

That said, I personally believe these two approaches — ORCDF’s structure-aware smoothing prevention and ICDM’s inductive graph-based reasoning — are highly complementary. It would be very interesting to explore combining the strengths of both. You’re welcome to give it a try!

You're most welcome to continue asking questions — I’d be glad to assist further.

**Comment by Holmes2002** [comment]
@lswhim Hi, I would like to ask if you have ever applied this model to a real online platform for Education. I have been struggling to find a Concept Competency assessment solution to find out which Concepts are suitable for students (not too easy and not too difficult). This model can solve my problem efficiently ? Can you give me some advice?

**Comment by lswhim** [comment]
Hi @Holmes2002 , thank you for reaching out with your question. Could you please provide more details about the Concept Competency assessment problem you are facing? From what I understand, it seems like you are looking for a solution to recommend suitable questions to students based on their previous performance. Is that correct? Providing more context will help us determine if this model can effectively address your needs.

**Comment by Holmes2002** [comment]
@lswhim Well, I have to build an educational recommendation system for students based on the ZPD (Zone of Proximal Development) principle — students should practice with exercises that are neither too easy nor too difficult to help them maximize their learning progress.

To tackle this problem, I divided it into two stages:
Stage 1: Based on the student's history, I evaluate their proficiency in different concepts, then filter out the ones where their scores are in the medium range (i.e., within their ZPD).
Stage 2: After identifying suitable concepts, I use knowledge tracing to predict which specific questions the student is likely to solve correctly and then recommend those questions.

I have already completed Stage 2, but I’m still struggling with Stage 1 — determining which concepts are appropriate for each student. Currently, I don't have much data. Even if I did, I’m unsure how much data would be sufficient for such an inductive model to work well. Also, I’m wondering whether a transductive model would be more suitable for a real-world educational platform, where the system might need to retrain frequently as new student data comes in.

**Comment by lswhim** [comment]
@Holmes2002 From your description, I believe transductive models like KaNCD or ORCDF might be more suitable choices for your task, where you need to recommend exercises that are neither too easy nor too difficult to help students maximize their learning progress. Transductive models are particularly useful for real-world educational platforms where the system might need to retrain frequently as new student data comes in. While ICDM is more suitable for scenarios requiring rapid diagnosis for a large number of new students (within one minute), you can utilize KaNCD or ORCDF to assess students' proficiency in different concepts. Then, as you mentioned, filter out the ones where their scores are in the medium range (i.e., within their ZPD). The training data would be the students' answer records, and the filtering criteria can be informed by the transductive model, which can also provide the difficulty level of the questions. You can evaluate whether the difficulty of a question exceeds the student's proficiency in the concept by too much to make your decision.

If you need further assistance or have more questions, feel free to ask!

**Comment by Holmes2002** [comment]
@lswhim Thanks for your advices !

**Comment by Holmes2002** [comment]
@lswhim Hi I want to ask you another question that in CDM models how can I decide that the student has completed the concept? Is it ok to set a threshold like knowledge tracking for question recommendation like >0.8? I not only want to estimate proficiency of concepts, but also when should annouce student has already completed this concept

**Comment by lswhim** [comment]
@Holmes2002 In CDM models, determining whether a student has mastered a concept using a fixed threshold can be challenging because it is highly dependent on the data and the characteristics of the student population. This makes it susceptible to group influences. Simply using the magnitude of the ability value to judge whether a student has mastered a concept is not always reasonable. This is because, in CDM models, the prediction function (or interaction function) inputs the difference between student ability and question difficulty. Therefore, in theory, the student's ability only needs to be greater than the question difficulty to predict correctly, making the ability value a relative concept rather than an absolute one.

To determine if a student has fully mastered a concept, you should use the obtained ability value to predict the accuracy of the student on all questions related to that concept. Then, set a threshold for this predicted accuracy. For example, you might decide that a predicted accuracy of over 0.8 (or 80%) indicates mastery. This approach accounts for the variability in question difficulty and the relative nature of ability scores, providing a more realistic assessment of a student's mastery of a concept.

If you need further assistance or have more questions, feel free to ask!

**Comment by Holmes2002** [comment]
@lswhim What does you mean the question difficulty? It is calculated = 1 - next question prediction score or other embedding like Mas ?

**Comment by lswhim** [comment]
@Holmes2002 In CDMs, the formula derived from traditional item response theory, which is expressed as a function ( f(\text{proficiency} - \text{difficulty}) ). In this context, ( f ) is represented by a neural network. Both question difficulty and student proficiency generally come from embeddings. This allows the model to assess how well a student's proficiency matches the difficulty of a question.

If you need further assistance or have more questions, feel free to ask!

**Comment by Holmes2002** [comment]
@lswhim Thanks ! It is clearly all my problem !

**Comment by Holmes2002** [comment]
@lswhim Hi I have another question that CDM models like KaNCD always have to retrain whenever there is a new student but in case an existing student in the database has new log responses. Do these models also have to retrain?

**Comment by lswhim** [comment]
@Holmes2002 For transductive models, whether you need to retrain depends on whether the new log responses come from new questions. If the new responses are from questions that were already included in the training set, retraining is not need. However, if the responses are from new questions, retraining would be required.

If your scenario frequently involves new students or new questions, I recommend looking into our other works: [LRCD](https://github.com/ECNU-ILOG/LRCD) and [DFCD](https://github.com/BW297/DFCD). LRCD is suitable for all diagnostic scenarios, including new questions, new students, new concepts, cross-domains, and cross-dataset situations. DFCD, on the other hand, is specialized for inductive scenarios, such as new questions, new students, and new concepts.

**Comment by Holmes2002** [comment]
But, In KaNCD Embedding Mas is implicit trained, questions are just used as building relationship questions and KCs. After training, If a existing students has a new log response ( existing questions in database) how can this embeding adapt or other CDM models can solve ?

**Comment by Holmes2002** [comment]
@lswhim Hi, can u help me solve this question?

**Comment by lswhim** [comment]
@Holmes2002 In transductive models, if a new log response from an existing student corresponds to a question that has already been answered by other students, then the model has already been trained on that question. This means that the question's embedding has been learned during the training phase, based on the responses from other students. Therefore, there is no need to adapt the embedding for this new log response, as the model can use the existing learned embeddings to make predictions or analyze the new response.

**Comment by Holmes2002** [comment]
@lswhim What does you mean new log response can adjust current knowledge state of existing student ? How can do that ? in KaNCD excersices are used as difficulty and discrinimation and I don't see any method for excersices can change current embedding student knowledge ?

**Comment by lswhim** [comment]
@Holmes2002 I feel like I might have misunderstood your meaning. Are you asking how, if a student answers a question from the existing question bank, how their Mas can be updated?

**Comment by Holmes2002** [comment]
@lswhim  Yeah, Student's Mas is estimated in Step N but when he did several questions how to update his MAS in CDM model,  I think model has to retrain everytime right ?

**Comment by lswhim** [comment]
@Holmes2002 If you want to obtain a more accurate and updated Mas for the student, you indeed need to retrain the model. However, if the number of new responses is small, you might only need to train on this new data. This incremental training can help update the student's knowledge state without having to retrain the entire model from scratch.

**Comment by Holmes2002** [comment]
@lswhim Great. I think I have no more questions about Edu-RecSys. Thanks for your helpful tips so far they have helped me a lot!
